### DISCUSSION TITLE
How to set up a SAML application to work with Databricks workspace SSO

### DISCUSSION THREAD

**Discussion by Rumbles** [zitadel/zitadel#9049] [Q&A]
I am trying to set up a Databricks workspace with Zitadel as the IDP, as described here: https://docs.databricks.com/en/archive/single-sign-on/index.html#verify-that-the-saml-response-is-signed

I created a SAML application in zitadel, initially letting it create the XML by providing it with the only URL databricks gives you for use in SSO:

`https://[YOUR_DATABRICKS_WORKSPACE]/saml/consume`

I provided this for both Entity ID and ACS URL, and using the XML it generated I attempted to login, but during the login I was redirected to a URL:

`https://[YOUR_DATABRICKS_WORKSPACE]/login.html?error=saml_response_not_signed_error`

I have tried a number of things to fix this, but so far I cannot find the solution.

I updated the metadata XML for my application so that the `SPSSODescriptor` was as follows:

`<md:SPSSODescriptor AuthnRequestsSigned="true" WantAssertionsSigned="true" ProtocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">`

I also added a section like this:

```
        <md:KeyDescriptor use="signing">
            <ds:KeyInfo>
                <ds:X509Data>
                    <ds:X509Certificate>${certificate_body}</ds:X509Certificate>
                </ds:X509Data>
            </ds:KeyInfo>
        </md:KeyDescriptor>
```

The certificate body comes from 

`https://[YOUR_ZITADEL_INSTANCE]/saml/v2/certificate`

I removed the header and footer and put it all on one line, I am able to copy the string from the Web UI and cat it in to ` | base64 -d | openssl x509 -text` and it appears to return the certificate. So I believe the cert is correct.

However, when I try to log in to databricks, I now get directed to a URL like:

`https://[YOUR_ZITADEL_INSTANCE]/saml/v2/SSO?SAMLRequest=pZJBb%2BIwEIX%2FSuS7HRICCRYgUehukSighu6hl8qxnWKR2MFjs7v%2Fft2wq7aH9rLHGc97nu9ppsDapqML7476QZ69BBf9ahsNtH%2BYIW81NQwUUM1aCdRxWi7uNzQlA9pZ4ww3DXon%2BVrBAKR1ymgUrVcztNvebnbf19tnkbE8y8ccF6OswhlPa1xULMeC58VkICdDkY9Q9ENaCNoZClbBAMDLtQbHtAutQZrhJMVJckjGNB3SwfgJRavAozRzveroXAc0jlmAJZ2vGsWJkBeidG2ZC5PkVEConNTueCbctPErUXxJ47LcoWj%2FF%2FdGaaH0y9ek1XUI6N3hsMf7XXlA0eIf%2FdJo8K20pbQXxeXjw%2BZtO1FxPEzYeFIUOS7EpCa8MV4QwRyrrOIneNuMX33QfPpa0j4RO%2F8Pp2n83mh6PY5tgFuv9ibk9Tv6ZmzL3OfsCUn6jhK47kep19BJrmolRYigaczPpZUh7xly1ksUxfPrrx%2BvcP4H&RelayState=%3Fhash%3D&returnurl=%3Fhash%3D`

And in my web browser I see:

```
<Response xmlns="urn:oasis:names:tc:SAML:2.0:protocol" ID="_e0137cba-2a82-43ed-977b-42f748feedaa" InResponseTo="ONELOGIN_d4a7476c-854b-4c2f-8ba7-dc7890e93d75" Version="2.0" IssueInstant="2024-12-11T16:23:06.667Z">
<Issuer xmlns="urn:oasis:names:tc:SAML:2.0:assertion" Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">https://[YOUR_ZITADEL_INSTANCE]/saml/v2/metadata</Issuer>
<Status xmlns="urn:oasis:names:tc:SAML:2.0:protocol">
<StatusCode xmlns="urn:oasis:names:tc:SAML:2.0:protocol" Value="urn:oasis:names:tc:SAML:2.0:status:RequestDenied"/>
<StatusMessage>failed to verify signature: %!w(<nil>)</StatusMessage>
</Status>
<Assertion xmlns="urn:oasis:names:tc:SAML:2.0:assertion" Version="" ID="" IssueInstant="">
<Issuer xmlns="urn:oasis:names:tc:SAML:2.0:assertion"/>
</Assertion>
</Response>
```

I then see an error in the logs on the zitadel server like:

```
zitadel-8696b49bc9-rmvkx zitadel time="2024-12-11T16:42:14Z" level=error msg="no signature provided but required" caller="/home/runner/go/pkg/mod/github.com/zitadel/saml@v0.2.0/pkg/provider/checker/checker.go:98"
```

I cannot see a way to configure zitadel to allow this situation and databricks doesn't appear to offer any way to configure to send a signature (or it's doing it in a way that zitadel doesn't recognise?)

Can anyone offer some more steps to try to understand this problem and resolve it please?

**Comment by ThomasSchoeller** [comment]
i had an similar issue. There is an patch in https://github.com/zitadel/saml/pull/102 to sign the full message which fixed my problem.

**Comment by avznog** [comment]
Hello, It happenned to me as well, does the patch works ?

**Comment by ThomasSchoeller** [comment]
Hey @avznog,

for my application it worked. Since [Version 4.7](https://github.com/zitadel/zitadel/releases/tag/v4.7.0) the saml patch is integrated in upstream. 

I think the issue can be closed with this as an solution.
BR
Thomas

**Reply by avznog** [reply]
Ok, thank you for your answer. I still am in 2.X in zitadel... will need to upgrade,
But thx !
